File: Sources/UsefulVoiceCore/Transcription/DeepgramProvider.swift
import Foundation

/// Deepgram Nova-3 pre-recorded transcription.
/// POST https://api.deepgram.com/v1/listen with the raw WAV bytes as the body.
/// @unchecked Sendable: no mutable state; both stored properties are Sendable.
public final class DeepgramProvider: TranscriptionProvider, @unchecked Sendable {
    public struct Config: Sendable {
        public let apiKey: String
        public let model: String
        public let smartFormat: Bool

        public init(apiKey: String, model: String = "nova-3", smartFormat: Bool) {
            self.apiKey = apiKey
            self.model = model
            self.smartFormat = smartFormat
        }
    }

    public let name = "Deepgram"
    private let config: Config
    private let session: URLSession

    public init(config: Config, session: URLSession = .shared) {
        self.config = config
        self.session = session
    }

    static let baseURL = URL(string: "https://api.deepgram.com/v1/listen")!

    /// Deadline floor for one attempt.
    static let minimumDeadline: TimeInterval = 15

    /// Absolute ceiling for one attempt, so a wedged request can never hang the
    /// pipeline forever.
    static let maximumDeadline: TimeInterval = 180

    /// Worst-case upload throughput assumed when sizing the deadline, in bytes
    /// per second (~0.8 Mbit/s). Deliberately pessimistic: dictation happens on
    /// tethered connections and hotel wifi, and the cost of being generous is
    /// only that a genuinely dead network takes longer to report.
    static let assumedUploadBytesPerSecond: Double = 100_000

    /// Fixed allowance for TLS setup, the server's own processing of Nova-3 on
    /// pre-recorded audio, and the response download.
    static let processingAllowance: TimeInterval = 12

    /// Wall-clock deadline for one attempt, sized to the audio being uploaded.
    ///
    /// A fixed 15 s cap covered upload *and* server processing, so a long
    /// dictation could never succeed: 16 kHz mono WAV is ~32 kB/s, which is
    /// ~9.6 MB for a five-minute recording — over 15 s of upload alone on a
    /// 5 Mbit/s link, before Deepgram has even looked at the audio. The app
    /// advertises recordings up to ten minutes, so the deadline has to scale.
    static func deadline(forAudioBytes bytes: Int) -> TimeInterval {
        let upload = Double(max(0, bytes)) / assumedUploadBytesPerSecond
        let total = upload + processingAllowance
        return min(maximumDeadline, max(minimumDeadline, total))
    }

    /// Nova-3 language codes. Auto-detect maps to `multi` (multilingual
    /// code-switching, a Nova-3 feature).
    static func languageParameter(for pin: LanguagePin) -> String {
        switch pin {
        case .en: return "en"
        case .de: return "de"
        case .auto: return "multi"
        }
    }

    public func makeRequest(audio: Data, hint: TranscriptionHint) throws -> URLRequest {
        let apiKey = config.apiKey.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !apiKey.isEmpty else {
            throw ProviderError.notConfigured("Enter your Deepgram API key.")
        }
        var components = URLComponents(url: Self.baseURL, resolvingAgainstBaseURL: false)!
        var items = [
            URLQueryItem(name: "model", value: config.model),
            URLQueryItem(name: "language", value: Self.languageParameter(for: hint.languagePin)),
        ]
        if config.smartFormat {
            items.append(URLQueryItem(name: "smart_format", value: "true"))
        }
        // Nova-3 keyterm prompting biases recognition toward the user's
        // vocabulary. Repeated once per term. The caller bounds this list: the
        // API rejects the whole request when the terms exceed its token limit,
        // and every term is URL-encoded into the request line.
        for term in hint.dictionaryWords where KeytermBudget.isSendableKeyterm(term) {
            items.append(URLQueryItem(name: "keyterm", value: term))
        }
        components.queryItems = items

        var request = URLRequest(url: components.url!)
        request.httpMethod = "POST"
        request.setValue("Token \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("audio/wav", forHTTPHeaderField: "Content-Type")
        request.httpBody = audio
        // Idle timeout only (it resets per byte); the real cap is the wall-clock
        // deadline below.
        request.timeoutInterval = Self.deadline(forAudioBytes: audio.count)
        return request
    }

    /// Races work against a true wall-clock deadline. URLRequest.timeoutInterval
    /// is only an idle timeout (it resets on every byte), so a slow trickling
    /// upload could exceed it indefinitely; this enforces a real total cap.
    ///
    /// Cancellation matters here: `group.cancelAll()` alone would leave the
    /// losing URLSession task running, so a timed-out request would keep
    /// uploading and billing. The work closure is wrapped on a task that is
    /// cancelled explicitly and awaited, which cancels the session task.
    static func withDeadline<T: Sendable>(
        seconds: TimeInterval,
        _ work: @escaping @Sendable () async throws -> T
    ) async throws -> T {
        let workTask = Task<T, Error> { try await work() }
        let timeoutTask = Task<Void, Never> {
            try? await Task.sleep(nanoseconds: UInt64(seconds * 1_000_000_000))
            workTask.cancel()
        }
        defer { timeoutTask.cancel() }

        do {
            return try await workTask.value
        } catch is CancellationError {
            throw ProviderError.timedOut
        } catch let urlError as URLError where urlError.code == .cancelled {
            throw ProviderError.timedOut
        }
    }

    public func transcribe(audio: URL, hint: TranscriptionHint) async throws -> Transcript {
        // Read the bytes off the caller's actor: this is file I/O and would
        // otherwise block the main thread inside the dictation pipeline.
        let audioData: Data
        do {
            audioData = try Data(contentsOf: audio, options: .mappedIfSafe)
        } catch {
            throw ProviderError.transport(URLError(.fileDoesNotExist))
        }
        let request = try makeRequest(audio: audioData, hint: hint)
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await Self.withDeadline(
                seconds: Self.deadline(forAudioBytes: audioData.count)
            ) { [session] in
                try await session.data(for: request)
            }
        } catch let urlError as URLError where urlError.code == .timedOut {
            throw ProviderError.timedOut
        } catch let urlError as URLError {
            throw ProviderError.transport(urlError)
        }
        guard let http = response as? HTTPURLResponse else { throw ProviderError.badResponse }
        guard (200..<300).contains(http.statusCode) else {
            // Surface a rate-limit hint rather than a bare 429.
            if http.statusCode == 429 {
                let retry = http.value(forHTTPHeaderField: "Retry-After")
                    .map { " Retry after \($0)s." } ?? ""
                throw ProviderError.http(429, "Rate limited by Deepgram.\(retry)")
            }
            throw ProviderError.http(http.statusCode, String(decoding: data, as: UTF8.self))
        }
        return try Self.parse(data)
    }

    static func parse(_ data: Data) throws -> Transcript {
        struct Response: Decodable {
            struct Metadata: Decodable { let duration: Double? }
            struct Results: Decodable {
                struct Channel: Decodable {
                    struct Alternative: Decodable { let transcript: String }
                    let alternatives: [Alternative]
                }
                let channels: [Channel]
            }
            let metadata: Metadata?
            let results: Results
        }
        guard let decoded = try? JSONDecoder().decode(Response.self, from: data),
              let transcript = decoded.results.channels.first?.alternatives.first?.transcript else {
            throw ProviderError.badResponse
        }
        return Transcript(
            text: transcript.trimmingCharacters(in: .whitespacesAndNewlines),
            detectedLanguage: nil,
            durationSeconds: decoded.metadata?.duration)
    }
}

File: windows/src/core/transcription/deepgramProvider.ts
/**
 * Deepgram Nova-3 pre-recorded transcription.
 *
 * Kept free of Electron so the request construction and error classification can
 * be unit-tested against a stubbed fetch.
 * https://developers.deepgram.com/reference/speech-to-text/listen-pre-recorded
 */

export interface DeepgramConfig {
  apiKey: string;
  smartFormat: boolean;
  /** Overrides the default endpoint; used by tests. */
  endpoint?: string;
}

export interface TranscriptionHint {
  language: string;
  /** At most a few hundred entries; always bounded by the keyterm budget. */
  keyterms: string[];
}

export interface Transcript {
  text: string;
  /** Provider-reported duration, when present. */
  durationSeconds: number | null;
  detectedLanguage: string | null;
}

export type ProviderErrorKind =
  | 'unauthorized'
  | 'badRequest'
  | 'rateLimited'
  | 'serverError'
  | 'timedOut'
  | 'transport'
  | 'cancelled'
  | 'malformedResponse';

export class ProviderError extends Error {
  readonly kind: ProviderErrorKind;
  readonly status?: number;
  /** Seconds to wait, from `Retry-After`, when the server supplied it. */
  readonly retryAfterSeconds?: number;

  constructor(
    kind: ProviderErrorKind,
    message: string,
    options: { status?: number; retryAfterSeconds?: number } = {},
  ) {
    super(message);
    this.name = 'ProviderError';
    this.kind = kind;
    this.status = options.status;
    this.retryAfterSeconds = options.retryAfterSeconds;
  }

  /**
   * Whether retrying the same request could plausibly succeed.
   *
   * A wrong API key or malformed audio will fail identically forever, so retrying
   * only makes the user wait longer for the same error.
   */
  get isTransient(): boolean {
    return (
      this.kind === 'rateLimited' ||
      this.kind === 'serverError' ||
      this.kind === 'timedOut' ||
      this.kind === 'transport'
    );
  }
}

export const DEFAULT_ENDPOINT = 'https://api.deepgram.com/v1/listen';

/**
 * Deadline components.
 *
 * A fixed total deadline cannot work for this app: recording is capped at 10
 * minutes, which at 16 kHz mono 16-bit is 19.2 MB. Delivering that inside a
 * fixed 15 seconds needs ~10 Mbps of sustained uplink, and the same window also
 * has to cover the TLS handshake and the server decoding a 10-minute file. The
 * result was that the advertised long-dictation feature could never succeed, and
 * the retry re-uploaded the same payload to time out again.
 */
export const MINIMUM_DEADLINE_SECONDS = 15;
/**
 * Hard ceiling on one attempt.
 *
 * Sized so that even the worst case this app can produce — a 10-minute dictation
 * (19.2 MB at 16 kHz mono 16-bit) uploading at the pessimistic assumed rate, plus
 * the processing allowance — is covered rather than cut off. A lower ceiling would
 * reintroduce exactly the bug this replaced: the advertised long-dictation
 * feature failing because the deadline assumed a fast connection.
 */
export const MAXIMUM_DEADLINE_SECONDS = 195;
/**
 * Upload throughput floor, in bytes per second, assumed when sizing the deadline.
 * ~100 KB/s is a pessimistic mobile/congested figure, so the deadline is generous
 * on any real connection.
 */
export const ASSUMED_UPLOAD_BYTES_PER_SECOND = 100_000;
/** Extra time for the server to decode the audio once it has all arrived. */
export const PROCESSING_ALLOWANCE_SECONDS = 12;

/**
 * Wall-clock deadline for one attempt at a given payload size.
 *
 * Rounded up to whole seconds because it is passed to `AbortSignal.timeout`,
 * which is millisecond-precision but read in logs as a human figure.
 */
export function deadlineForAudioBytes(bytes: number): number {
  const uploadSeconds = Math.max(0, bytes) / ASSUMED_UPLOAD_BYTES_PER_SECOND;
  const total = uploadSeconds + PROCESSING_ALLOWANCE_SECONDS;
  return Math.min(
    MAXIMUM_DEADLINE_SECONDS,
    Math.max(MINIMUM_DEADLINE_SECONDS, Math.ceil(total)),
  );
}

export interface BuildRequestOptions {
  audioBytes: number;
  hint: TranscriptionHint;
  config: DeepgramConfig;
}

export interface BuiltRequest {
  url: string;
  headers: Record<string, string>;
  deadlineSeconds: number;
}

/** Build the Deepgram request (URL, headers, deadline). */
export function buildRequest({ audioBytes, hint, config }: BuildRequestOptions): BuiltRequest {
  const endpoint = config.endpoint ?? DEFAULT_ENDPOINT;
  const params = new URLSearchParams();
  params.set('model', 'nova-3');
  // `multi` is the documented auto-detect mode for Nova-3.
  params.set('language', hint.language === 'auto' ? 'multi' : hint.language);
  params.set('smart_format', config.smartFormat ? 'true' : 'false');
  params.set('punctuate', 'true');

  // `keyterm` is Nova-3 only and repeats once per term. The list must already be
  // inside the token ceiling; this filter is a second line of defence so a
  // malformed entry can never reach the request line.
  for (const keyterm of hint.keyterms) {
    if (keyterm.trim().length > 0) params.append('keyterm', keyterm.trim());
  }

  return {
    url: `${endpoint}?${params.toString()}`,
    headers: {
      Authorization: `Token ${config.apiKey}`,
      'Content-Type': 'audio/wav',
    },
    deadlineSeconds: deadlineForAudioBytes(audioBytes),
  };
}

interface DeepgramAlternative {
  transcript?: string;
}

interface DeepgramChannel {
  alternatives?: DeepgramAlternative[];
  detected_language?: string;
}

interface DeepgramResponse {
  results?: {
    channels?: DeepgramChannel[];
    duration?: number;
  };
  metadata?: { duration?: number };
}

/** Extract the transcript from a Deepgram pre-recorded response. */
export function parseResponse(payload: unknown): Transcript {
  if (typeof payload !== 'object' || payload === null) {
    throw new ProviderError('malformedResponse', 'The transcription service returned an unexpected response.');
  }
  const body = payload as DeepgramResponse;
  const channel = body.results?.channels?.[0];
  const text = channel?.alternatives?.[0]?.transcript ?? '';
  const duration =
    body.results?.duration ??
    body.metadata?.duration ??
    null;
  return {
    text: typeof text === 'string' ? text.trim() : '',
    durationSeconds: typeof duration === 'number' && Number.isFinite(duration) ? duration : null,
    detectedLanguage: channel?.detected_language ?? null,
  };
}

/**
 * Short, safe description of an error body.
 *
 * Response bodies can be large and can echo the request; only a bounded prefix is
 * kept, and the API key is redacted if it ever appears.
 */
export function describeErrorBody(body: string, apiKey?: string): string {
  let text = body.slice(0, 300).replace(/\s+/g, ' ').trim();
  if (apiKey && apiKey.length > 0) {
    text = text.split(apiKey).join('[redacted]');
  }
  return text;
}

/**
 * Classify a non-2xx response.
 *
 * Transient failures (429, 5xx) are separated from permanent ones (401, 400) so
 * the caller can retry only what can actually succeed.
 */
export function classifyHttpError(status: number, body: string, apiKey?: string): ProviderError {
  const detail = describeErrorBody(body, apiKey);
  const suffix = detail.length > 0 ? `: ${detail}` : '';
  if (status === 401 || status === 403) {
    return new ProviderError('unauthorized', `The Deepgram API key was rejected${suffix}`, { status });
  }
  if (status === 429) {
    return new ProviderError('rateLimited', `Deepgram rate limit reached${suffix}`, { status });
  }
  if (status >= 500) {
    return new ProviderError('serverError', `Deepgram is having trouble (HTTP ${status})${suffix}`, { status });
  }
  return new ProviderError('badRequest', `Transcription failed (HTTP ${status})${suffix}`, { status });
}

/** Parse a `Retry-After` header, which may be seconds or an HTTP date. */
export function parseRetryAfter(value: string | null, now: Date = new Date()): number | undefined {
  if (!value) return undefined;
  const seconds = Number(value.trim());
  if (Number.isFinite(seconds) && seconds >= 0) return Math.min(seconds, 60);
  const date = Date.parse(value);
  if (Number.isNaN(date)) return undefined;
  return Math.min(Math.max(0, (date - now.getTime()) / 1000), 60);
}

export interface TranscribeOptions {
  audio: Uint8Array;
  hint: TranscriptionHint;
  config: DeepgramConfig;
  /** Injected for tests. */
  fetchImpl?: typeof fetch;
  /** Abort signal from the caller (user pressed Escape). */
  signal?: AbortSignal;
  /** Attempts for transient failures. 1 disables retrying. */
  maxAttempts?: number;
  /** Injected for tests, so the suite never really sleeps. */
  sleep?: (ms: number) => Promise<void>;
}

export const MAX_ATTEMPTS = 3;

/**
 * Transcribe a WAV buffer.
 *
 * Retries transient failures with exponential backoff and honours `Retry-After`.
 * A single momentary network blip used to fail the user's dictation outright and
 * demand a manual retry.
 */
export async function transcribe(options: TranscribeOptions): Promise<Transcript> {
  const { audio, hint, config } = options;
  const doFetch = options.fetchImpl ?? fetch;
  const sleep = options.sleep ?? defaultSleep;
  const maxAttempts = Math.max(1, options.maxAttempts ?? MAX_ATTEMPTS);

  const request = buildRequest({ audioBytes: audio.byteLength, hint, config });

  let lastError: ProviderError | undefined;
  for (let attempt = 1; attempt <= maxAttempts; attempt += 1) {
    if (options.signal?.aborted) {
      throw new ProviderError('cancelled', 'Transcription was cancelled.');
    }
    try {
      return await attemptOnce(request, audio, doFetch, options.signal);
    } catch (error) {
      const providerError = toProviderError(error);
      lastError = providerError;
      if (!providerError.isTransient || attempt === maxAttempts) throw providerError;
      // Cap the server's own guidance so a large Retry-After cannot stall the
      // user's dictation for minutes.
      const delaySeconds = Math.min(providerError.retryAfterSeconds ?? 0.5 * 3 ** (attempt - 1), 8);
      await sleep(delaySeconds * 1000);
    }
  }
  throw lastError ?? new ProviderError('transport', 'Transcription failed.');
}

async function attemptOnce(
  request: BuiltRequest,
  audio: Uint8Array,
  doFetch: typeof fetch,
  signal: AbortSignal | undefined,
): Promise<Transcript> {
  const timeoutSignal = AbortSignal.timeout(request.deadlineSeconds * 1000);
  const combined = signal ? AbortSignal.any([signal, timeoutSignal]) : timeoutSignal;

  let response: Response;
  try {
    response = await doFetch(request.url, {
      method: 'POST',
      headers: request.headers,
      // A copy, because fetch in Node rejects a SharedArrayBuffer-backed view and
      // some runtimes mutate the buffer.
      body: new Uint8Array(audio),
      signal: combined,
    });
  } catch (error) {
    if (signal?.aborted) {
      throw new ProviderError('cancelled', 'Transcription was cancelled.');
    }
    if (isAbortError(error)) {
      throw new ProviderError(
        'timedOut',
        `Transcription timed out after ${request.deadlineSeconds}s.`,
      );
    }
    throw new ProviderError(
      'transport',
      `Could not reach Deepgram: ${error instanceof Error ? error.message : String(error)}`,
    );
  }

  if (!response.ok) {
    const body = await safeText(response);
    const error = classifyHttpError(response.status, body);
    if (error.kind === 'rateLimited') {
      const retryAfter = parseRetryAfter(response.headers.get('retry-after'));
      throw new ProviderError('rateLimited', error.message, {
        status: response.status,
        ...(retryAfter !== undefined ? { retryAfterSeconds: retryAfter } : {}),
      });
    }
    throw error;
  }

  let payload: unknown;
  try {
    payload = await response.json();
  } catch {
    throw new ProviderError('malformedResponse', 'The transcription service returned invalid JSON.');
  }
  return parseResponse(payload);
}

function isAbortError(error: unknown): boolean {
  if (error instanceof DOMException && error.name === 'TimeoutError') return true;
  if (error instanceof Error) {
    return error.name === 'AbortError' || error.name === 'TimeoutError';
  }
  return false;
}

function toProviderError(error: unknown): ProviderError {
  if (error instanceof ProviderError) return error;
  return new ProviderError(
    'transport',
    error instanceof Error ? error.message : String(error),
  );
}

async function safeText(response: Response): Promise<string> {
  try {
    return await response.text();
  } catch {
    return '';
  }
}

function defaultSleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
